ipld/retriever_quadrant.go

```diff
@@ -0,0 +1,102 @@
+package ipld
+
+import (
+	"math"
+	"math/rand"
+
+	"github.com/ipfs/go-cid"
+	"github.com/tendermint/tendermint/pkg/da"
+
+	"github.com/celestiaorg/celestia-node/ipld/plugin"
+)
+
+// there are always 4 quadrants
+const numQuadrants = 4
+
+type quadrant struct {
+	// slice of roots to get shares from
+	roots []cid.Cid
+	// Example coordinates(x;y) of each quadrant when fetching from column roots
+	// ------  -------
+	// |  Q0 | | Q1  |
+	// |(0;0)| |(1;0)|
+	// ------  -------
+	// |  Q2 | | Q3  |
+	// |(0;1)| |(1;1)|
+	// ------  -------
+	x, y int
+	// source defines the axis for quadrant
+	// it can be either 1 or 0 similar to x and y
+	// where 0 is Row source and 1 is Col respectively
+	source int
+}
+
+// newQuadrants constructs a slice of quadrants from DAHeader.
+// Constantly, there are 4 quadrants for each source: row and col (8 in total).
+// The ordering of quadrants is random.
+func newQuadrants(dah *da.DataAvailabilityHeader) [][]*quadrant {
```

Comment:
can you add a unit test for this?

Reply:
Will add in a subsequent PR as we discussed.